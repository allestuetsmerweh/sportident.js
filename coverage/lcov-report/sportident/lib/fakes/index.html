
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for sportident/lib/fakes</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> sportident/lib/fakes</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">29.1% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>39/134</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">34.4% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>43/125</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">18.51% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>5/27</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">28.94% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>33/114</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="FakeSiMainStation.js"><a href="FakeSiMainStation.js.html">FakeSiMainStation.js</a></td>
	<td data-value="17.07" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 17%"></div><div class="cover-empty" style="width: 83%"></div></div>
	</td>
	<td data-value="17.07" class="pct low">17.07%</td>
	<td data-value="82" class="abs low">14/82</td>
	<td data-value="21.21" class="pct low">21.21%</td>
	<td data-value="66" class="abs low">14/66</td>
	<td data-value="6.66" class="pct low">6.66%</td>
	<td data-value="15" class="abs low">1/15</td>
	<td data-value="17.33" class="pct low">17.33%</td>
	<td data-value="75" class="abs low">13/75</td>
	</tr>

<tr>
	<td class="file low" data-value="IFakeSiMainStation.js"><a href="IFakeSiMainStation.js.html">IFakeSiMainStation.js</a></td>
	<td data-value="29.03" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 29%"></div><div class="cover-empty" style="width: 71%"></div></div>
	</td>
	<td data-value="29.03" class="pct low">29.03%</td>
	<td data-value="31" class="abs low">9/31</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="35" class="abs low">14/35</td>
	<td data-value="14.28" class="pct low">14.28%</td>
	<td data-value="7" class="abs low">1/7</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="24" class="abs low">8/24</td>
	</tr>

<tr>
	<td class="file medium" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="76.19" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 76%"></div><div class="cover-empty" style="width: 24%"></div></div>
	</td>
	<td data-value="76.19" class="pct medium">76.19%</td>
	<td data-value="21" class="abs medium">16/21</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="24" class="abs medium">15/24</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="15" class="abs high">12/15</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-01-27T22:19:42.487Z
            </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    