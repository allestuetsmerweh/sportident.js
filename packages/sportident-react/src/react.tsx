import React from 'react';
import * as Immutable from 'immutable';
import {SiDeviceAddEvent, ISiDeviceDriverWithAutodetection, SiDeviceRemoveEvent} from 'sportident/lib/SiDevice/ISiDeviceDriver';
import {ISiDevice, ISiDeviceDriverData} from 'sportident/lib/SiDevice/ISiDevice';

export const useSiDevices = <T extends ISiDeviceDriverData<unknown>>(
    siDeviceDriver: ISiDeviceDriverWithAutodetection<T>,
): Immutable.Map<string, ISiDevice<T>> => {
    const [siDevices, setSiDevices] = React.useState<Immutable.Map<string, ISiDevice<T>>>(Immutable.Map({}));
    React.useEffect(() => {
        const onDeviceAdd = (event: SiDeviceAddEvent<T>) => {
            const device = event.siDevice;
            setSiDevices((currentSiDevices) => {
                if (!currentSiDevices.has(device.ident)) {
                    console.log('useSiDevices: add');
                    return currentSiDevices.set(device.ident, device);
                }
                return currentSiDevices;
            });
        };
        const onDeviceRemove = (event: SiDeviceRemoveEvent<T>) => {
            const device = event.siDevice;
            setSiDevices((currentSiDevices) => {
                if (currentSiDevices.has(device.ident)) {
                    console.log('useSiDevices: remove');
                    return currentSiDevices.delete(device.ident);
                }
                return currentSiDevices;
            });
        };
        siDeviceDriver.addEventListener('add', onDeviceAdd);
        siDeviceDriver.addEventListener('remove', onDeviceRemove);
        siDeviceDriver.startAutoDetection().then((devices) => {
            setSiDevices((currentSiDevices) => {
                const existingIdentSet = Immutable.Set.fromKeys(currentSiDevices);
                const newIdentSet = Immutable.Set(devices.map((device) => device.ident));
                if (!newIdentSet.equals(existingIdentSet)) {
                    console.log('useSiDevices: reset');
                    return Immutable.Map(devices.map((device) => [device.ident, device]));
                }
                return currentSiDevices;
            });
        });
        console.log('useSiDevices: setup');
        return () => {
            console.log('useSiDevices: cleanup');
            siDeviceDriver.stopAutoDetection();
            siDeviceDriver.removeEventListener('add', onDeviceAdd);
            siDeviceDriver.removeEventListener('remove', onDeviceRemove);
        };
    }, []);
    return siDevices as Immutable.Map<string, ISiDevice<T>>;
};
