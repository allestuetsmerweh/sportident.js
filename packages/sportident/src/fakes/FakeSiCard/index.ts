export {BaseFakeSiCard} from './BaseFakeSiCard';
export * from './types';
