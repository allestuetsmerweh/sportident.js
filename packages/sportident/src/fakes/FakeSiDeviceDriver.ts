import * as utils from '../utils';
import {DeviceClosedError, ISiDevice, ISiDeviceDriverData, SiDeviceState} from '../SiDevice/ISiDevice';
import {ISiDeviceDriver, ISiDeviceDriverWithAutodetection, ISiDeviceDriverWithDetection, SiDeviceDriverWithAutodetectionEvents, SiDeviceAddEvent, SiDeviceRemoveEvent} from '../SiDevice/ISiDeviceDriver';
import {SiDevice} from '../SiDevice/SiDevice';

interface FakeDevice {
    ident: string;
    isOpened: boolean;
}

const getIdent = (fakeDevice: FakeDevice): string => fakeDevice.ident;

export interface FakeSiDeviceDriverData extends ISiDeviceDriverData<FakeSiDeviceDriver> {
    driver: FakeSiDeviceDriver;
    device: FakeDevice;
}

export type IFakeSiDevice = ISiDevice<FakeSiDeviceDriverData>;
export type FakeSiDevice = SiDevice<FakeSiDeviceDriverData>;

// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
class FakeSiDeviceDriver implements
        ISiDeviceDriver<FakeSiDeviceDriverData>,
        ISiDeviceDriverWithDetection<FakeSiDeviceDriverData, []>,
        ISiDeviceDriverWithAutodetection<FakeSiDeviceDriverData> {
    public name = 'Fake';

    private siDeviceByIdent:
        {[ident: string]: FakeSiDevice} = {};

    private autodetectedSiDevices:
        {[ident: string]: FakeSiDevice} = {};

    private autoDetectionRunning = false;

    public identToBeDetected: string|null = null;
    public identsToBeAutodetected: string[] = [];

    detect(): Promise<FakeSiDevice> {
        if (!this.identToBeDetected) {
            return Promise.reject(new Error('No fake device ident to be detected'));
        }
        return this.autodetectSiDevice({
            ident: this.identToBeDetected,
            isOpened: false,
        });
    }

    getSiDevice(fakeDevice: FakeDevice): FakeSiDevice {
        const ident = getIdent(fakeDevice);
        if (this.siDeviceByIdent[ident] !== undefined) {
            return this.siDeviceByIdent[ident];
        }
        const newSiDeviceData: FakeSiDeviceDriverData = {
            driver: this,
            device: fakeDevice,
        };
        const newSiDevice = new SiDevice(ident, newSiDeviceData);
        this.siDeviceByIdent[ident] = newSiDevice;
        return newSiDevice;
    }

    async forgetSiDevice(siDevice: FakeSiDevice): Promise<void> {
        const fakeDevice = siDevice.data.device;
        const ident = getIdent(fakeDevice);
        delete this.siDeviceByIdent[ident];
        if (this.autodetectedSiDevices[ident] !== undefined) {
            await this.dispatchEvent('remove', new SiDeviceRemoveEvent(siDevice));
        }
        delete this.autodetectedSiDevices[ident];
    }

    public isAutoDetectionRunning(): boolean {
        return this.autoDetectionRunning;
    }

    startAutoDetection(): Promise<IFakeSiDevice[]> {
        this.autoDetectionRunning = true;
        return this.getAutodetectedDevices();
    }

    getAutodetectedDevices(): Promise<FakeSiDevice[]> {
        return this.autodetectSiDevices(this.identsToBeAutodetected);
    }

    autodetectSiDevices(idents: string[]): Promise<FakeSiDevice[]> {
        return Promise.all(
            idents.map((ident) => this.autodetectSiDevice({ident, isOpened: false})),
        );
    }

    autodetectSiDevice(fakeDevice: FakeDevice): Promise<FakeSiDevice> {
        const ident = getIdent(fakeDevice);
        if (this.autodetectedSiDevices[ident] !== undefined) {
            return Promise.reject(new Error('Duplicate SI device'));
        }
        const siDevice = this.getSiDevice(fakeDevice);
        this.autodetectedSiDevices[ident] = siDevice;
        return Promise.resolve(siDevice);
    }

    public async handleDeviceConnected(ident: string): Promise<void> {
        if (!this.autoDetectionRunning) {
            throw new Error('autodetection not running');
        }
        const openedDevice = await this.autodetectSiDevice({ident, isOpened: false});
        await this.dispatchEvent('add', new SiDeviceAddEvent(openedDevice));
    }

    public async handleDeviceDisconnected(ident: string): Promise<void> {
        if (!this.autoDetectionRunning) {
            throw new Error('autodetection not running');
        }
        const siDevice = this.siDeviceByIdent[ident];
        if (siDevice === undefined) {
            throw new Error('No such device');
        }
        await this.forgetSiDevice(siDevice);
    }

    stopAutoDetection(): Promise<unknown> {
        this.autoDetectionRunning = false;
        return Promise.resolve();
    }

    open(device: IFakeSiDevice): Promise<void> {
        device.data.device.isOpened = true;
        return Promise.resolve();
    }

    close(device: IFakeSiDevice): Promise<unknown> {
        device.data.device.isOpened = false;
        return Promise.resolve();
    }

    receive(device: IFakeSiDevice): Promise<number[]> {
        const fakeDevice = device.data.device;
        if (fakeDevice.isOpened !== true) {
            device.setState(SiDeviceState.Closed);
            throw new DeviceClosedError();
        }
        return Promise.resolve([]);
    }

    send(_device: IFakeSiDevice, _uint8Data: number[]): Promise<unknown> {
        return Promise.resolve(true);
    }
}
// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging, @typescript-eslint/no-empty-object-type
interface FakeSiDeviceDriver extends utils.EventTarget<SiDeviceDriverWithAutodetectionEvents<FakeSiDeviceDriverData>> {}
utils.applyMixins(FakeSiDeviceDriver, [utils.EventTarget]);

export const getFakeSiDeviceDriver = (): FakeSiDeviceDriver => new FakeSiDeviceDriver();
