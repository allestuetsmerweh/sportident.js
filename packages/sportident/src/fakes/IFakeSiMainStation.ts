import * as utils from '../utils';
import * as siProtocol from '../siProtocol';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface IFakeSiMainStation {}

export class FakeSiMainStationMessageEvent extends utils.Event<'message'> {
    constructor(
                public fakeSiMainStation: IFakeSiMainStation,
                public message: siProtocol.SiMessage,
    ) {
        super();
    }
}

export type FakeSiMainStationEvents = {
    'message': FakeSiMainStationMessageEvent,
};
