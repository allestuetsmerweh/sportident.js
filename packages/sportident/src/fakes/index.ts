export {FakeSiMainStation} from './FakeSiMainStation';
export * from './FakeSiCard';
