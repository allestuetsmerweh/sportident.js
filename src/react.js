import React from 'react';
import Immutable from 'immutable';

export const useSiDevices = (siDeviceClass) => {
    const [siDevices, setSiDevices] = React.useState(Immutable.Map({}));
    React.useEffect(() => {
        const onDeviceAdd = (event) => {
            const device = event.webUsbSiDevice;
            if (!siDevices.has(device.ident)) {
                const newSiDevices = siDevices.set(device.ident, device);
                console.log('add');
                setSiDevices(newSiDevices);
            }
        };
        const onDeviceRemove = (event) => {
            const device = event.webUsbSiDevice;
            if (siDevices.has(device.ident)) {
                const newSiDevices = siDevices.delete(device.ident);
                console.log('remove');
                setSiDevices(newSiDevices);
            }
        };
        siDeviceClass.addEventListener('add', onDeviceAdd);
        siDeviceClass.addEventListener('remove', onDeviceRemove);
        siDeviceClass.startAutoDetection().then((devices) => {
            const existingIdentSet = Immutable.Set.fromKeys(siDevices);
            const newIdentSet = Immutable.Set(devices.map((device) => device.ident));
            if (!newIdentSet.equals(existingIdentSet)) {
                console.log('reset');
                const newSiDevices = Immutable.Map(devices.map((device) => [device.ident, device]));
                setSiDevices(newSiDevices);
            }
        });
        console.log('setup');
        return () => {
            console.log('cleanup');
            siDeviceClass.stopAutoDetection();
            siDeviceClass.removeEventListener('add', onDeviceAdd);
            siDeviceClass.removeEventListener('remove', onDeviceRemove);
        };
    }, [siDevices]);
    return siDevices;
};
