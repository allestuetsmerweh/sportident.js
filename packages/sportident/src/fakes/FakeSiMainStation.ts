import {proto} from '../constants';
import * as utils from '../utils';
// eslint-disable-next-line no-unused-vars
import * as storage from '../storage';
import * as siProtocol from '../siProtocol';
// eslint-disable-next-line no-unused-vars
import {ISiStationStorageFields, siStationStorageDefinition} from '../SiStation/BaseSiStation';
// eslint-disable-next-line no-unused-vars
import {FakeSiMainStationEvents, FakeSiMainStationMessageEvent} from './IFakeSiMainStation';
// eslint-disable-next-line no-unused-vars
import {IFakeSiCard} from './FakeSiCard/IFakeSiCard';

export class FakeSiMainStation {
    public storage: storage.ISiStorage<ISiStationStorageFields>;
    public isMaster = true;
    public dateOffset = 0;
    public fakeSiCard?: IFakeSiCard;

    constructor(storageArg: (number|undefined)[]|undefined) {
        this.storage = siStationStorageDefinition(storageArg);
    }

    dispatchCardMessage(cardMessage: siProtocol.SiMessage) {
        if (cardMessage.mode !== undefined) {
            return;
        }
        const message = {
            command: cardMessage.command,
            parameters: [...this.getCode(), ...cardMessage.parameters],
        };
        this.dispatchMessage(message);
    }

    dispatchMessage(message: siProtocol.SiMessage) {
        this.dispatchEvent(
            'message',
            new FakeSiMainStationMessageEvent(this, message),
        );
    }

    getCode(): number[] {
        return [
            ((this.storage.data.get(0x73)! & 0xC0) << 2),
            this.storage.data.get(0x72)!,
        ];
    }

    getDateTime(): Date {
        return new Date(Date.now() + this.dateOffset);
    }

    insertCard(fakeSiCard: IFakeSiCard) {
        this.fakeSiCard = fakeSiCard;
        const cardMessage = this.fakeSiCard.handleDetect();
        this.dispatchCardMessage(cardMessage);
    }

    sendMessage(message: siProtocol.SiMessage) {
        if (message.mode !== undefined) {
            return;
        }
        if (message.command === proto.cmd.SIGNAL) {
            const numSignal = message.parameters[0];
            this.dispatchMessage({
                command: proto.cmd.SIGNAL,
                parameters: [...this.getCode(), numSignal],
            });
        } else if (message.command === proto.cmd.GET_MS) {
            this.dispatchMessage({
                command: proto.cmd.GET_MS,
                parameters: [...this.getCode(), this.isMaster ? proto.P_MS_DIRECT : proto.P_MS_REMOTE],
            });
        } else if (message.command === proto.cmd.SET_MS) {
            this.isMaster = message.parameters[0] === proto.P_MS_DIRECT;
            this.dispatchMessage({
                command: proto.cmd.SET_MS,
                parameters: [...this.getCode(), this.isMaster ? proto.P_MS_DIRECT : proto.P_MS_REMOTE],
            });
        } else if (message.command === proto.cmd.GET_TIME) {
            this.dispatchMessage({
                command: proto.cmd.GET_TIME,
                parameters: [...this.getCode(), ...siProtocol.date2arr(this.getDateTime())],
            });
        } else if (message.command === proto.cmd.SET_TIME) {
            const newTime = siProtocol.arr2date(message.parameters.slice(0, 7));
            console.log(newTime); // TODO: Use new time to set dateOffset
            this.dispatchMessage({
                command: proto.cmd.SET_TIME,
                parameters: [...this.getCode(), ...siProtocol.date2arr(this.getDateTime())],
            });
        } else if (message.command === proto.cmd.GET_SYS_VAL) {
            const offset = message.parameters[0];
            const length = message.parameters[1];
            this.dispatchMessage({
                command: proto.cmd.GET_SYS_VAL,
                parameters: [
                    ...this.getCode(),
                    offset,
                    ...this.storage.data.slice(offset, offset + length),
                ] as number[],
            });
        } else if (message.command === proto.cmd.SET_SYS_VAL) {
            const offset = message.parameters[0];
            const newContent = message.parameters.slice(1);
            let data = this.storage.data;
            newContent.forEach((newByte: number, index: number) => {
                data = data.set(offset + index, newByte);
            });
            this.storage = siStationStorageDefinition(data);
            this.dispatchMessage({
                command: proto.cmd.SET_SYS_VAL,
                parameters: [...this.getCode(), offset],
            });
        } else if (
            message.command === proto.cmd.GET_SI5
            || message.command === proto.cmd.GET_SI6
            || message.command === proto.cmd.GET_SI8
        ) {
            if (this.fakeSiCard === undefined) {
                return;
            }
            const cardMessages = this.fakeSiCard.handleRequest(message);
            cardMessages.forEach((cardMessage: siProtocol.SiMessage) => {
                this.dispatchCardMessage(cardMessage);
            });
        } else if (message.command === proto.cmd.ERASE_BDATA) {
            this.dispatchMessage({
                command: proto.cmd.ERASE_BDATA,
                parameters: [...this.getCode()],
            });
        }
    }
}
// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface FakeSiMainStation extends utils.EventTarget<FakeSiMainStationEvents> {}
utils.applyMixins(FakeSiMainStation, [utils.EventTarget]);
