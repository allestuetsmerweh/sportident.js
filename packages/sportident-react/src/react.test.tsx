import {describe, expect, test} from '@jest/globals';
import {render} from '@testing-library/react';
import _ from 'lodash';
import React from 'react';
import {getFakeSiDeviceDriver} from 'sportident/lib/fakes/FakeSiDeviceDriver';
import * as react from './react';

describe('react', () => {
    test('exists', () => {
        expect(react).not.toBe(undefined);
    });
    test('works', async () => {
        const fakeSiDeviceDriver = getFakeSiDeviceDriver();
        fakeSiDeviceDriver.identsToBeAutodetected = ['auto-1'];
        const SiDevicesList = () => {
            const fakeSiDevices = react.useSiDevices(fakeSiDeviceDriver);

            const deviceList = [...fakeSiDevices.values()].map((device) => (
                <div key={`device-${device.ident}`}>
                    {device.name} ({device.ident})
                </div>
            ));

            return (
                <div>
                    {deviceList}
                </div>
            );
        };

        const {container, rerender, unmount} = render(<SiDevicesList/>);

        expect(container).toMatchSnapshot();
        expect(fakeSiDeviceDriver.isAutoDetectionRunning()).toEqual(true);

        // Wait until initial autodetection
        await Promise.resolve();
        await Promise.resolve();
        rerender(<SiDevicesList/>);

        expect(container).toMatchSnapshot();
        expect(fakeSiDeviceDriver.isAutoDetectionRunning()).toEqual(true);

        // Autodetect new device
        await fakeSiDeviceDriver.handleDeviceConnected('auto-2');
        rerender(<SiDevicesList/>);

        expect(container).toMatchSnapshot();
        expect(fakeSiDeviceDriver.isAutoDetectionRunning()).toEqual(true);

        // Remove autodetected device again
        await fakeSiDeviceDriver.handleDeviceDisconnected('auto-2');
        rerender(<SiDevicesList/>);

        expect(container).toMatchSnapshot();
        expect(fakeSiDeviceDriver.isAutoDetectionRunning()).toEqual(true);

        unmount();

        expect(container).toMatchSnapshot();
        expect(fakeSiDeviceDriver.isAutoDetectionRunning()).toEqual(false);
    });
});
