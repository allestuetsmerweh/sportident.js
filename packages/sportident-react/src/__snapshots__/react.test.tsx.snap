// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react works 1`] = `
<div>
  <div />
</div>
`;

exports[`react works 2`] = `
<div>
  <div>
    <div>
      Fake(auto-1)
       (
      Fake-auto-1
      )
    </div>
  </div>
</div>
`;

exports[`react works 3`] = `
<div>
  <div>
    <div>
      Fake(auto-1)
       (
      Fake-auto-1
      )
    </div>
    <div>
      Fake(auto-2)
       (
      Fake-auto-2
      )
    </div>
  </div>
</div>
`;

exports[`react works 4`] = `
<div>
  <div>
    <div>
      Fake(auto-1)
       (
      Fake-auto-1
      )
    </div>
  </div>
</div>
`;

exports[`react works 5`] = `<div />`;
